src/CrossChainReceiverFactory.sol

```diff
@@ -0,0 +1,644 @@
+            proof.length := calldataload(proof.offset)
+            proof.offset := add(0x20, proof.offset)
+            extraData.offset := add(signature.offset, calldataload(add(0x40, signature.offset)))
+            extraData.length := calldataload(extraData.offset)
+            extraData.offset := add(0x20, extraData.offset)
+        }
+
+        if (extraData.length != 0) {
+            _verifyRoot(MerkleProofLib.getRoot(proof, ""), owner);
+            if (_erc1271IsValidSignature(hash, extraData, owner))
+                return IERC1271.isValidSignature.selector;
+            return 0xffffffff;
+        }
+
+        _verifyRoot(MerkleProofLib.getRoot(proof, hash), owner);
+        return IERC1271.isValidSignature.selector;
+    }
+
+    function eip712Domain()
+        public
+        view
+        onlyProxy
+        returns (
+            bytes1 fields,
+            string memory name_,
+            string memory,
+            uint256 chainId,
+            address verifyingContract,
+            bytes32,
+            uint256[] memory
+        )
+    {
+        fields = bytes1(0x0d);
+        name_ = name;
+        chainId = block.chainid;
+        verifyingContract = address(this);
+    }
+
+    function deploy(bytes32 root, address owner, bool setOwner) external noDelegateCall returns (address proxy) {
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            // derive the deployment salt from the owner and chainid
+            mstore(0x14, owner)
+            mstore(0x00, root)
+            let salt := keccak256(0x00, 0x34)
+
+            // create a minimal proxy targeting this contract
+            mstore(0x1d, 0x5af43d5f5f3e6022573d5ffd5b3d5ff3)
+            mstore(0x0d, address())
+            mstore(0x00, 0x60265f8160095f39f35f5f365f5f37365f6c)
+            proxy := create2(0x00, 0x0e, 0x2f, salt)
+            if iszero(proxy) {
+                mstore(0x00, 0x30116425) // selector for `DeploymentFailed()`.
+                revert(0x1c, 0x04)
+            }
+
+            // restore clobbered memory
+            mstore(0x40, ptr)
+
+            // If `setOwner == true`, this gets the selector for `setPendingOwner(address)`,
+            // otherwise you get the selector for `cleanup(address)`. In both cases, the selector is
+            // appended with `owner`'s padding
+            let selector := xor(0xfbacefce000000000000000000000000, mul(0x3f8c8622000000000000000000000000, setOwner))
+
+            // set the pending owner, or `selfdestruct` to the owner
+            mstore(0x14, owner)
+            mstore(0x00, selector)
+            if iszero(call(gas(), proxy, 0x00, 0x10, 0x24, 0x00, 0x00)) {
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+        }
+    }
+
+    function setPendingOwner(address owner) external onlyFactory {
+        _setPendingOwner(owner);
+    }
```

Comment:
let's replace this (and the above selector trick) with `setOwner` rather than `setPendingOwner` because I think that this is introducing pointless additional friction and complexity